### PR TITLE
Associated functions that contain extern indicator or have `#[rustc_std_internal_symbol]` are reachable

### PR DESCRIPTION
**Pull request by ghost**
Previously these fails to link with ``undefined reference to `foo'``:

<details>
<summary>Example 1</summary>

```rs
struct AssocFn;

impl AssocFn {
    #[no_mangle]
    fn foo() {}
}

fn main() {
    extern "Rust" {
        fn foo();
    }
    unsafe { foo() }
}
```
([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=f1244afcdd26e2a28445f6e82ca46b50))
</details>

<details>
<summary>Example 2</summary>

```rs
#![crate_name = "lib"]
#![crate_type = "lib"]

struct AssocFn;

impl AssocFn {
    #[no_mangle]
    fn foo() {}
}
```
```rs
extern crate lib;

fn main() {
    extern "Rust" {
        fn foo();
    }
    unsafe { foo() }
}
```
</details>

But I believe they should link successfully, because this works:
<details>

```rs
#[no_mangle]
fn foo() {}

fn main() {
    extern "Rust" {
        fn foo();
    }
    unsafe { foo() }
}
```
([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=789b3f283ee6126f53939429103ed98d))
</details>

This PR fixes the problem, by adding associated functions that have "custom linkage" to `reachable_set`, just like normal functions.

I haven't tested whether #76211 and [Miri](https://github.com/rust-lang/miri/issues/1837) are fixed by this PR yet, but I'm submitting this anyway since this fixes the examples above.

I added a `run-pass` test that combines my two examples above, but I'm not sure if that's the right way to test this. Maybe I should add / modify an existing codegen test (`src/test/codegen/export-no-mangle.rs`?) instead?

### CODE REVIEW

**Comment by petrochenkov** [comment]
I don't think `#[no_mangle]` is supposed to work with associated functions at all.
Are there practical use cases / uses in the wild?

All the checks and logic mentioning `sym::no_mangle` that I can find in the compiler only apply to free items.

**Comment by petrochenkov** [comment]
I'm still not sure whether the label combination `S-waiting-on-team` + `T-lang` is enough for the lang team to triage a PR, so I'm going to nominate as well.

**Comment by hyd-dev** [comment]
> I don't think `#[no_mangle]` is supposed to work with associated functions at all.
> Are there practical use cases / uses in the wild?

I don't know. I find it by reading this code:
https://github.com/rust-lang/rust/blob/2962e7c0089d5c136f4e9600b7abccfbbde4973d/compiler/rustc_codegen_ssa/src/back/symbol_export.rs#L84-L89

**Comment by Mark-Simulacrum** [comment]
This was discussed during the last several language team meetings. There doesn't seem to be a lot of strong opinions in either direction on this, but the overall consensus seems to be that we should make `#[no_mangle]` work (i.e., produce a non-mangled symbol in the final binary, along with export/reachability) on functions inside `impl Foo`. This seems to be consistent with a general perspective that the context where a function is placed should not, generally, restrict what attributes can be placed on it so long as they are supportable in that context.

There was some debate over whether this is practically useful, but generally it seems like the answer is at least *maybe* yes - the description of https://github.com/rust-lang/rust/issues/76211 cites a concrete case where this was unexpectedly not working, for example. Ultimately, it seems relatively harmless to support this, and fairly likely to do what users expect.

----

We did not discuss whether it makes sense to feature gate this to start -- I suspect that would in practice break some code in the wild, and this seems largely like a bugfix, so perhaps we will opt to skip that. In that case we'll want an FCP so I'll bring this up on Zulip so that someone on T-lang can hopefully kick that off.

**Comment by nikomatsakis** [comment]
@rfcbot fcp merge

Per @Mark-Simulacrum's [summary](https://github.com/rust-lang/rust/pull/86492#issuecomment-885682960), I propose to merge this. I think we should skip the feature gate, as I consider this a bug fix.

**Comment by cramertj** [comment]
Sharing in case others have the same thought process:

At first I thought that it was odd to have a `#[no_mangle]` associated function use the `fn <...>` name, since I think of the name of those functions as including the type-component of the path. That is, I think of the function above as `AssocFn::foo` rather than a function named `foo`. However, the more I consider it, the more it makes sense to think of it just like we would `mod assoc_fn { #[no_mangle] fn foo() { ... } }`-- the type is part of the path / item container, just as any other path element such as a crate or module would be.

**Comment by petrochenkov** [comment]
If the intent is to support `#[no_mangle]` on associated items, then this PR is not enough.
You can search for `sym::no_mangle` in rustc codebase - all those places need to be updated to account for `no_mangle` on non-items.

**Comment by hyd-dev** [comment]
I've adjusted every `sym::no_mangle` and `sym::export_name` check I've found. This includes extending the "`#[no_mangle]` requires ASCII identifier" error to associated functions, which could be breaking.

I think many changes I made have effects on functions in `impl Trait for Foo { ... }` in addition to `impl Foo { ... }` as well. I'm not sure whether this is covered by the lang team decision.

@rustbot label -S-waiting-on-author S-waiting-on-review

**Comment by petrochenkov** [comment]
Looks like the `check_no_mangle` check in `compiler\rustc_passes\src\check_attr.rs` also allows `#[no_mangle]` attributes on trait items like this:
```rust
trait Tr {
    #[no_mangle] fn method() {} // with or without body
}
```
Apparently `#[no_mangle]` should affect the symbol name for the default method implementation as well, if it is generated.
We need to either support it properly on trait methods with bodies, or just prohibit it everywhere except free functions.
(Besides that `check_no_mangle` should certainly be modified to warn on `#[no_mangle]` applied to functions without body.)

That's why I was skeptical about supporting this from the start, it results in some amount of (largely useless) feature creep.

**Comment by petrochenkov** [comment]
@hyd-dev 
Could you then warn on them in `check_no_mangle`?

**Comment by hyd-dev** [comment]
Done.

I also changed `check_export_name` to report errors for `#[export_name = "..."]` on trait methods.

The warning is `UNUSED_ATTRIBUTES`, but I think the attribute is actually used: https://godbolt.org/z/hcaMbj1c1. Maybe we can completely [ignore the attribute](https://github.com/hyd-dev/rust/commit/44dae2e977d7d6f94a5d30ba683a90771cb661c5).

@rustbot label -S-waiting-on-author S-waiting-on-review

**Comment by nbdd0121** [comment]
`#[no_mangle]` definitely shouldn't be allowed on trait methods; it wouldn't work if you instantiate two copies: https://godbolt.org/z/sqbGnfx5M.

**Comment by petrochenkov** [comment]
https://github.com/hyd-dev/rust/commit/44dae2e977d7d6f94a5d30ba683a90771cb661c5 looks like the right thing to do to me (assuming `#[no_mangle]` in all generic cases produces a warning).

By the way, does `#[no_mangle]` produce a warning in cases like this:
```rust
impl<T> Foo<T> { // <- generic parameter is here
    #[no_mangle] fn foo() {} // <- no generic parameter here, but the function may be instantiated multiple times
}
```

**Comment by petrochenkov** [comment]
That said, `#[no_mangle]` does work on all generic functions right now, including default trait methods, it only produces duplicate symbol errors if the function is actually instantiated twice.
Perhaps this behavior is intentional and useful in some cases? Otherwise `no_mangle_generic_items` would be an error and not a lint.

**Comment by hyd-dev** [comment]
> By the way, does `#[no_mangle]` produce a warning in cases like this:

Yes, it does: https://github.com/rust-lang/rust/pull/86492/files#diff-1da490505bfd4472e9487a7341d7e0a11b8af3fd5fadc6156ec1c3f0da28ba10R81-R83

> Otherwise `no_mangle_generic_items` would be an error and not a lint.

The lint was added in #30269. It seems that `#[no_mangle]` is "[sliently ignored](https://github.com/rust-lang/rust/issues/15844)" at that time.
I'm not sure whether the current behaviour could be useful, though.

**Comment by hyd-dev** [comment]
It seems there's no equivalent lint for `#[export_name = "..."]`:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b6ae83cb4cd0e90a879f771f1aeefdda

I'm not sure if that's intentional or not ([I didn't find any issue/PR related to that](https://github.com/rust-lang/rust/issues?q=export_name+generic)), but this PR makes `#[export_name = "..."]` on a trait method produce an error: https://github.com/rust-lang/rust/pull/86492/files#diff-3a4de9e66643a61b5125db1e6618eb95f9cbe993bd09b45f772fddc3751604cdR113-R115.

---

> Can you add a test case which checks that

Done.

**Comment by hyd-dev** [comment]
It seems there's no equivalent lint for `#[export_name = "..."]`:
https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=b6ae83cb4cd0e90a879f771f1aeefdda

I'm not sure if that's intentional or not ([I didn't find any issue/PR related to that](https://github.com/rust-lang/rust/issues?q=export_name+generic)), but this PR makes `#[export_name = "..."]` on a trait method produce an error: https://github.com/rust-lang/rust/pull/86492/files#diff-3a4de9e66643a61b5125db1e6618eb95f9cbe993bd09b45f772fddc3751604cdR113-R115.

---

> Can you add a test case which checks that

Done.

**Comment by hyd-dev** [comment]
CI failure is https://rust-lang.zulipchat.com/#narrow/stream/242791-t-infra/topic/mingw-check.20broken.

File: compiler/rustc_passes/src/reachable.rs [hunk 1]
```diff
@@ -211,13 +211,15 @@ impl<'tcx> ReachableContext<'tcx> {
...
-                let codegen_attrs = self.tcx.codegen_fn_attrs(item.def_id);
+            if let Node::Item(hir::Item { kind: hir::ItemKind::Fn(sig, ..), def_id, .. })
+            | Node::ImplItem(hir::ImplItem {
+                kind: hir::ImplItemKind::Fn(sig, ..),
+                def_id,
+                ..
+            }) = *node
+            {
+                let reachable = sig.header.abi != Abi::Rust;
+                let codegen_attrs = self.tcx.codegen_fn_attrs(*def_id);
```
Comment (hyd-dev) [inline compiler/rustc_passes/src/reachable.rs:222]: I think this change (unintentionally) made `reachable_set` no longer include any `#[no_mangle]` static item. Opened #88032 to fix this.

**Pull request merged**